### PR TITLE
Add a Decider for generational tuning

### PR DESCRIPTION
**Pull request by sidheart**
- This PR will be broken down into smaller chunks at a later date
- Add metrics into the YoungGenRca to detect undersized young
generations
- Add a YoungGenDecisionPolicy to suggest modification actions when an
issue is detected
- Add a GenerationalSizingDecider to make generation tuning decisions
- Modify the NodeConfigCollector so that NewRatio can be determined
dynamically on the ES Cluster
- Add a PersistableSlidingWindow class to allow large sliding windows to
be persisted and loaded from disk to avoid losing data on node restarts

*Fixes #:*

*Description of changes:*

*Tests:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  @Override
+  public void next(SlidingWindowData slidingWindowData) {
+    if (timestamp == -1) {
+      timestamp = slidingWindowData.getTimeStamp();
+    }
+    long currTimestamp = slidingWindowData.getTimeStamp();
+    super.next(slidingWindowData);
+    long timestampDiff = currTimestamp - timestamp;
+    if (timeUnit.convert(timestampDiff, TimeUnit.MILLISECONDS)
+        > timeUnit.convert(writePeriod, TimeUnit.MILLISECONDS)) {
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:58]: Why are we doing this? Are you trying to ensure at most one entry in queue per hour? If so, then shouldn't `super.next()` be called within this `if` ? and outside the if we should just be incrementing a count within that data point (if at all needed).

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java [hunk 2]
```diff
@@ -53,6 +55,9 @@
 
   private static final Logger LOG = LogManager.getLogger(HighHeapUsageYoungGenRca.class);
   private static final int PROMOTION_RATE_SLIDING_WINDOW_IN_MINS = 10;
+  private static final int ONE_HOUR_IN_MIN = 60;
+  private static final double GARBAGE_PROMOTION_THRESHOLD = .80;
+  private static final double FOLLOWER_CHECK_TIMEOUT_MS = 30d * 1_000;
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:60]: why is this called `FOLLOWER_CHECK_*` ? is this related to leader/follower checks in es?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java [hunk 3]
```diff
@@ -118,6 +133,67 @@ protected void remove(SlidingWindowData e) {
     this(rcaPeriod, 1.0, heap_Used, gc_Collection_Time);
   }
 
+  private boolean fullGcTimeTooHigh(double avgFullGcTime) {
+    return (!Double.isNaN(avgFullGcTime) && avgFullGcTime > FOLLOWER_CHECK_TIMEOUT_MS);
+  }
+
+  private boolean promotionRateTooHigh(double avgPromotionRate, double modifier) {
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:140]: Minor: is the `modifier` intended to change the threshold? if so, can we just override the threshold in an overloaded fn call?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java [hunk 4]
```diff
@@ -118,6 +133,67 @@ protected void remove(SlidingWindowData e) {
...
+
+  private ResourceFlowUnit<HotResourceSummary> compute(double avgPromotionRate,
+                                                       double avgYoungGCTime,
+                                                       double avgGarbagePromoted,
+                                                       double avgFullGCTime) {
+    ResourceContext context = new ResourceContext(State.UNHEALTHY);
+    HotResourceSummary summary = null;
+    boolean unhealthy = true;
+
+    // Check if the RCA is unhealthy
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:166]: should this be an if-else-if? or should we let the same node indicate multiple issues - they can be a signal of urgency to upstream consumers

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java [hunk 5]
```diff
@@ -132,6 +208,22 @@ protected void remove(SlidingWindowData e) {
...
       if (!Double.isNaN(oldGenHeapUsed)) {
         promotionRateDeque.next(new SlidingWindowData(currTimeStamp, oldGenHeapUsed / CONVERT_BYTES_TO_MEGABYTES));
+        if (oldGenHeapUsed > maxOldGen) {
+          maxOldGen = oldGenHeapUsed;
+        }
+        double promoted = oldGenHeapUsed - prevOldGen;
+        if (promoted > 0) {
+          youngGenPromotedMB += promoted / CONVERT_BYTES_TO_MEGABYTES;
+        } else { // full GC must have occurred if there's less data in the old gen
+          double garbageReclaimedPct = (prevOldGen - oldGenHeapUsed) / (youngGenPromotedMB * CONVERT_BYTES_TO_MEGABYTES);
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:218]: `HighHeapUsageOldGenRca` has a `minOldGenSlidingWindow` that gives old gen just after gc. should we use that here? it does a bunch of things to ensure we get the min value of old-gen-usage if a gc happened.

also nit: `promoted == 0` does not imply a gc.

Comment (vigyasharma) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:218]: should be ok for rcas to consume each other, or we can do it in the cluster rca; or simply refactor

**Review by vigyasharma — COMMENTED** [review]
Thanks for sharing this draft PR @sidheart. Super useful in getting the approach out.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Action class is used to modify the cache's max size. It is used by cache decider and other
+ * deciders to implement actions like increasing the cache's size. Presently, it acts on field data
+ * cache and shard request cache.
+ */
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java:29]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+public class ModifyJvmGenerationAction extends SuppressibleAction {
+  public static final String NAME = "ModifyJvmGeneration";
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java:31]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    return canUpdate;
+  }
+
+  @Override
+  public long coolOffPeriodInMillis() {
+    return coolOffPeriodInMillis;
+  }
+
+  @Override
+  public List<NodeKey> impactedNodes() {
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyJvmGenerationAction.java:65]: Marked impacted nodes as data nodes, the actual implementation of this will be in CoS or the remediation plugin

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    return (int) Math.floor(currentRatio + 1);
+  }
+
+  public int increaseYoungGeneration() {
+    double currentRatio = getCurrentRatio();
+    // Don't increase the (old:young) ratio beyond 3:1
+    if (currentRatio < 3) {
+      return -1;
+    }
+    return (int) Math.floor(currentRatio - 1);
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java:106]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      if (newRatio >= 1) {
+        actions.add(new ModifyJvmGenerationAction(appContext, increaseYoungGeneration(), COOLOFF_PERIOD_IN_MILLIS, true));
+      }
+    }
+    return actions;
+  }
+
+  public void setAppContext(AppContext appContext) {
+    this.appContext = appContext;
+    this.issueCountThreshold = (int) Math.ceil(appContext.getDataNodeInstances().size() * .1d);
```
[the remark this answers is not in the document]
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java:135]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java (hunk 1 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java:58]: The way I had this before, I was writing every write period, but not bucketizing the data.

I've changed this so that data is bucketized and writes are tied to bucket width (in units of time)

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java (hunk 2 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:60]: The idea is that this is related to leader/follower checks. I noticed that in a lot of Sev2s, when this issue causes node drops it's because GC time starts approaching this value.

This may need to be dynamic eventually, but for now a value around the default for our clusters should be good enough

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java (hunk 3 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:140]: This is an artifact of a refactor of the operate() method. I only included the modifier param to be able to factor out this logic. Since this isn't related to code I wrote, I'm going to leave this as is for now.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java (hunk 4 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:166]: This is a good idea, but I'd rather implement this as a future iteration. For now, given the current aggregation strategy, a simple issue count is enough.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java (hunk 5 above)
Comment (sidheart) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:218]: You're right. I started factoring out a common RCA to do these computations, but I'm leaving it as a future iteration for now.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java [hunk 11]
```diff
@@ -51,20 +55,17 @@
       OLD_GEN_HEAP_USAGE
   );
   private static final long COOLOFF_PERIOD_IN_MILLIS = 48L * 24L * 60L * 60L * 1000L;
+  private static final String UNDERSIZED_DATA_FILE_PATH = "/tmp/JvmGenerationTuningPolicy_Undersized";
+  private static final String OVERSIZED_DATA_FILE_PATH = "/tmp/JvmGenerationTuningPolicy_Oversized";
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java:59]: nit: do we want these in `/tmp`? Is there another config or data folder that we can use?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/FileRotate.java [hunk 12]
```diff
@@ -126,43 +126,47 @@ protected synchronized Path rotate(long currentMillis) throws IOException {
...
         LOG.error("Could not delete file: " + targetFilePath);
       }
       try {
-        ret = Files.move(FILE_TO_ROTATE, targetFilePath, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+        ret = Files.move(fileToRotate, targetFilePath, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
       } catch (Exception ex) {
         LOG.error(ex);
-        LOG.error("Deleting file: {}", FILE_TO_ROTATE);
-        tryDelete(FILE_TO_ROTATE);
+        LOG.error("Deleting file: {}", fileToRotate);
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/persistence/FileRotate.java:157]: I hope this includes the exception stack trace, if not pls add.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java [hunk 13]
```diff
@@ -11,59 +26,79 @@
...
-    super.next(slidingWindowData);
-    long timestampDiff = currTimestamp - timestamp;
-    if (timeUnit.convert(timestampDiff, TimeUnit.MILLISECONDS)
-        > timeUnit.convert(writePeriod, TimeUnit.MILLISECONDS)) {
+    long timestampDiff = currTimestamp - aggregatedData.getTimeStamp();
+    if (timeUnit.convert(timestampDiff, TimeUnit.SECONDS) > bucketSizeInSeconds) {
       try {
+        super.next(aggregatedData);
+        aggregatedData = null;
         persist();
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java]: So we only persist when new buckets get added (at hour boundaries), not when individual bucket counters get updated?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java [hunk 14]
```diff
@@ -103,35 +104,52 @@ public int increaseYoungGeneration() {
...
     List<Action> actions = new ArrayList<>();
-    if (tooLargeIssues.size() > getOversizedThreshold()) {
+    if (tooLargeIssues.readCurrentBucket() > policyConfig.getOversizedbucketHeight()) {
       int newRatio = decreaseYoungGeneration();
       if (newRatio >= 1) {
-        actions.add(new ModifyJvmGenerationAction(appContext, decreaseYoungGeneration(), COOLOFF_PERIOD_IN_MILLIS, true));
+        actions.add(new ModifyJvmGenerationParams(appContext, decreaseYoungGeneration(), COOLOFF_PERIOD_IN_MILLIS, true));
       }
-    } else if (tooSmallIssues.size() > getUndersizedThreshold()) {
+    } else if (tooSmallIssues.readCurrentBucket() > policyConfig.getUndersizedbucketHeight()) {
```
Comment (vigyasharma) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/JvmGenerationTuningPolicy.java:139]: I'm a bit confused here. IIUC, `readCurrentBucket()` only returns the size of current hour bucket. If we only need to decide based on current hour values, then why are we saving the full sliding window?

**Comment by sidheart** [comment]
This PR has been broken up into various pieces. The decider piece is contained in #460
